### ISSUE TITLE
Overrides of the `__class__` attributes in a few places can cause issues with PyCharm debugging (and possibly other PEP669 related tracing)

### ISSUE THREAD

**Issue by yvan-the-barbarian** [DataDog/dd-trace-py#7950]
### Summary of problem
There are few places in code where the `__class__` attribute is overridden with a property or through `__getattribute__` in order to spoof the class of the object, but the implementation causes issues if the `__init__` method of the object hasn't been called yet.

Examples I've found so far:
* `_ImportHookChainedLoader` overrides `__getattribute__`:
  1. If `__class__` is requested, it will attempt to return `self.loader.__class__`.
  2. But if the object hasn't been fully initialized yet, `self.loader` does not exist yet.
  3. This causes `__getattr__` to be called, which gets the attribute of the same name (here `loader`) on the `self.loader` attribute to be returned. But `self.loader` does not exist yet.
  4. This causes an infinite loop and a stack overflow error.
* `ObjectProxy` overrides `__class__` using a property.
  1. If `__class__` is requested, it will attempt to return `self.__wrapped__.__class__`.
  2. But if the object hasn't been fully initialized yet,  `self.__wrapped__` does not exist yet.
  3. This causes `__getattr__` to be called, which has specific code which raises `ValueError` if `__wrapped__` is requested.

Except, there are legitimate reasons why something might check the class of an object before the object is fully initialized, causing a crash.

A very frustrating example right now is PyCharm, which at the start of each function call will check the type of the `self` variable (if any) in order to detect if the current frame is in it's own debugger code so it can cull that frame.

The instant a `_ImportHookChainedLoader` or a `ObjectProxy` is initialized, the debugger code will detect the call to `__init__`, check the local variables, find `self` and check its type with `isinstance`, this in turn fetches the value of `__class__`, which causes an exception to be raised because `__init__` hasn't had the chance to run yet!

Unfortunately, this means that trying to debug any project using Python 3.12 in a PyCharm environment is impossible as an exception is raised the instant `ddtrace` is imported.

While I've already submitted a bug report to JetBrains concerning PyCharm's debugging (suggesting that the code shouldn't assume that `isinstance` won't raise an exception), there may be other tracing utilities which might suffer from this problem. After all, having `isinstance` fail is still kind of unusual.

A way to mitigate these issues would be to have calls to the `__class__` attribute work even if `__init__` hasn't been called yet. Perhaps the "real" class should be returned until the class of the wrapped or spoofed object is available.

Basically, the litmus test would be that the following code would work without raising an exception:

```python
# Where "cls" is the class to test.
instance = cls.__new__(...)
isinstance(cls, object)
```

### Which version of dd-trace-py are you using?
Tested with versions 2.0.0 and 2.3.2

### How can we reproduce your problem?
1. Install `ddtrace` 2.0.0 or 2.3.2 in a virtual environment.
5. Create a project with the following code:
   ```python
   import ddtrace
   ```
6. Using PyCharm, run the file _in debug mode_.

### What is the result that you get?
Exception with message `ValueError: wrapper has not been initialized` is raised.

### What is the result that you expected?
No exception raised.
